### PR TITLE
implement SPI

### PR DESCRIPTION
**Pull request by strengejacke**
#521 

This does not yet change all the docs, I think this would result in too many conflicts, probably. So we might mention the "SPI" option in the docs later.

### CODE REVIEW

**Comment by bwiernik** [comment]
Do we want to copy the function from SPIn rather than soft-importing, considering that the package hasn't been updated since 2013?

**Comment by strengejacke** [comment]
I wasn't sure about that, but this can be done easily. We just need the code in `.spi()`. It's more a moral question to me about copying the code 1:1 ;-)

**Comment by strengejacke** [comment]
That method or the code doesn't seem to be very robust:

``` r
library(bayestestR)
x <- rnorm(1000)
spi(x) |> as.data.frame()
#>     CI    CI_low  CI_high
#> 1 0.95 -1.851012 2.079607
SPIn::SPIn(x)$spin
#> [1] -1.851012  2.079607

m <- insight::download_model("brms_1")
d <- as.data.frame(m)
sapply(d, spi)
#> Error in quadprog::solve.QP(D.u, d.u, A.u, c(1, rep(0, u.u - u.l + 2)), : constraints are inconsistent, no solution!

sapply(d, SPIn::SPIn)
#> Error in solve.QP(D.u, d.u, A.u, c(1, rep(0, u.u - u.l + 2)), u.u - u.l): constraints are inconsistent, no solution!
```

<sup>Created on 2022-04-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by bwiernik** [comment]
The constraints are the `A.l` matrix. I would need to look more into the code for how this is generated and the paper this is based on to understand what's going on

**Comment by DominiqueMakowski** [comment]
should I submit without this PR?

**Comment by DominiqueMakowski** [comment]
> should I submit without this PR or wait for a bit?

*bump* 😬

**Comment by strengejacke** [comment]
w/o this PR, I'd say we must first take a closer look and see how this function can be made more stable. The original `SPIn()` in the _SPIn_ package has some more arguments, which might help adjusting the algorithm to not fail in the above example - but it doesn't help user for whom it is probably still too opaque at the moment what they could do.

**Comment by strengejacke** [comment]
But you _could_ elaborate a bit on https://github.com/easystats/bayestestR/issues/519 :-)I

**Comment by bwiernik** [comment]
@strengejacke The issue that's causing it to fail is the bandwidth argument `bw` is somewhat too large for the intercept and lp_ parameters. 

Let's try this:
1. Use `tryCatch()` with the function using the default `bw = round((sqrt(n.sims) - 1)/2)`
2. If that errors, try again multiplying that bw by .9, .8, then .7

That should make it safe for most distributions.

**Comment by bwiernik** [comment]
(I'll take care of that)

**Comment by bwiernik** [comment]
```r
library(bayestestR)
m <- insight::download_model("brms_1")
d <- as.data.frame(m)
sapply(d, spi)
# > sapply(d, spi)
#         b_Intercept b_wt      b_cyl      sigma    lp__     
# CI      0.95        0.95      0.95       0.95     0.95     
# CI_low  36.25617    -4.703458 -2.386442  1.985231 -83.11161
# CI_high 43.31241    -1.567123 -0.6765242 3.385691 -77.98328
```

**Comment by bwiernik** [comment]
Assuming the checks pass, we can include this

**Comment by strengejacke** [comment]
@DominiqueMakowski have you already submitted? If yes, it's no big deal, but if not, we may include this PR as well - we still need to add this to the docs as well.

**Comment by strengejacke** [comment]
I don't know, this test seems to me like the results are a bit random...

https://github.com/easystats/bayestestR/runs/5856422169?check_suite_focus=true#step:10:336

**Comment by bwiernik** [comment]
That's seems very random. We might want to try to stabilize the seed on that test? But seems unrelated to the current changes

File: R/spi.R [hunk 1]
```diff
@@ -249,9 +253,10 @@ spi.get_predicted <- function(x, ...) {
 
 
   # lower bound
-  if (k == 1) {
+  if (!is.na(k) && all(k == 1)) {
     x1 <- l
     w.l <- 1
+    x.l <- NA
```
Comment (strengejacke) [inline R/spi.R:259]: @bwiernik There are some edge cases, where this if-statement is TRUE, and then `x.l` was not initialized. I set it to `NA`, is that ok?

**Comment by bwiernik** [comment]
What if statement do you mean?

**Comment by strengejacke** [comment]
We have

```r
  # upper bound
  if (!is.na(ui) && all(ui == n.sims)) {
    x2 <- u
    w.u <- 1
    x.u <- NA # <- this here is new
  } else {
```

and at the bottom of the function:

```r
  # output
  data.frame(
    "CI" = ci,
    "CI_low" = x.l,
    "CI_high" = x.u
  )
```

when `if (!is.na(ui) && all(ui == n.sims)) {` was TRUE, `x.u` wasn't initialized, so I added `x.u <- NA # <- this here is new` to the if-statement. I also did some minor fixes based on the tests I added, and fixed all errors/issues I found. Mostly rare edge cases, as you can see from the tests. But in general, it now works well with models.

**Comment by strengejacke** [comment]
See my last changes here: https://github.com/easystats/bayestestR/pull/522/commits/a22371d74e8e7544b4cd4ad0bfec7cff94e6af65
Based on the tests I copied from the test-hdi.R file. This showed some issues I fixed in that commit.

**Comment by strengejacke** [comment]
@IndrajeetPatil why does test-coverage always fail? Is there something wrong with the script/action?

**Comment by bwiernik** [comment]
Oh, no. The `x1` and `x2` variables assigned above should just be changed to `x.l` and `x.u`. Missed updating those

**Comment by strengejacke** [comment]
Cool, I think we're set.

**Comment by bwiernik** [comment]
Not sure what's up with those tests
